opentelemetry-api/src/opentelemetry/trace/__init__.py

```diff
@@ -261,10 +263,39 @@ class TraceState(typing.Dict[str, str]):
         https://www.w3.org/TR/trace-context/#tracestate-field
     """
 
+    MAX_TRACESTATE_VALUES = 32
+    KEY_WITHOUT_VENDOR_FORMAT = r"[a-z][_0-9a-z\-\*\/]{0,255}"
+    KEY_WITH_VENDOR_FORMAT = (
+        r"[a-z][_0-9a-z\-\*\/]{0,240}@[a-z][_0-9a-z\-\*\/]{0,13}"
+    )
+    KEY_FORMAT = KEY_WITHOUT_VENDOR_FORMAT + "|" + KEY_WITH_VENDOR_FORMAT
+    VALUE_FORMAT = (
+        r"[\x20-\x2b\x2d-\x3c\x3e-\x7e]{0,255}[\x21-\x2b\x2d-\x3c\x3e-\x7e]"
+    )
+
+    KEY_VALIDATION_RE = re.compile("^" + KEY_FORMAT + "$")
+    VALUE_VALIDATION_RE = re.compile("^" + VALUE_FORMAT + "$")
+
     @classmethod
     def get_default(cls) -> "TraceState":
         return cls()
 
+    def __setitem__(self, key: str, value: str) -> None:
+        # According to the w3c spec, we can only store 32 values
+        if len(self) >= self.MAX_TRACESTATE_VALUES:
+            return
```

Comment:
I guess we want to pop the oldest value and take the new one?

Reply:
Yeah, by this logic the tracestate becomes immutable once it reaches the limit. However, overwriting an existing key does not change length and should work IMO.

Also I am reading the limit in the spec as a MAY, so IMO it would make sense to apply the limit during extraction only (but not prevent people from sending large tracestate).

Reply:
IMHO we should just keep TraceState being a dictionary (or ordered dictionary) and not entangle it with the W3C spec more closely than required. So the formatter would just strip superfluous values on extract (and maybe inject).

Reply:
sounds good. @unitaker your point is valid: existing keys should be overrideable. Can add.

@reyang the spec is ambiguous as to what should be done when a maximum is reached. I think either could be an issue.

@Oberon00 that is a good point. will move it to the extract / injection part. Or probably remove that limit entirely.

All that said it's a good point that this is an optional choice. I'd probably just vote to remove this restriction in the basic implementation: I worry about people not knowing the spec well enough and running into a tracecontext member limit.

Reply:
@toumorokoshi the [spec](https://github.com/w3c/trace-context/blob/master/spec/20-http_header_format.md#tracestate-limits) is clear about this:
```
Then entries SHOULD be removed starting from the end of tracestate.
Note that other truncation strategies like safe list entries, blocked list entries,
or size-based truncation MAY be used, but are highly discouraged.
```